File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: list[int]) -> list[list[int]]:
        return permutations([], nums)

def permutations(p, up):
    if not up:
        return [p]

    ans = []

    ch = up[0]

    for i in range(len(p) + 1):
        f = p[:i]
        s = p[i:]
        ans += permutations(f + [ch] + s, up[1:])

    return ans

File: 0046-permutations/0046-permutations.java
import java.util.*;

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        ArrayList<Integer> arr = new ArrayList<>();

        for (int num : nums) {
            arr.add(num);
        }

        return permutations(new ArrayList<>(), arr);
    }

    static List<List<Integer>> permutations(List<Integer> p, List<Integer> up) {
        if (up.isEmpty()) {
            ArrayList<List<Integer>> list = new ArrayList<>();
            list.add(p);
            return list;
        }

        int ch = up.get(0);
        ArrayList<List<Integer>> ans = new ArrayList<>();

        for (int i = 0; i <= p.size(); i++) {
            List<Integer> f = new ArrayList<>(p.subList(0, i));
            List<Integer> s = new ArrayList<>(p.subList(i, p.size()));

            f.add(ch);
            f.addAll(s);

            List<Integer> remaining =
                    new ArrayList<>(up.subList(1, up.size()));

            ans.addAll(permutations(f, remaining));
        }

        return ans;
    }
}